README.md
# Docker Flighttracker

[![Build release images](https://github.com/Dennis14e/docker-flighttracker/actions/workflows/build-release.yml/badge.svg)](https://github.com/Dennis14e/docker-flighttracker/actions/workflows/build-release.yml)
[![Build develop images](https://github.com/Dennis14e/docker-flighttracker/actions/workflows/build-develop.yml/badge.svg)](https://github.com/Dennis14e/docker-flighttracker/actions/workflows/build-develop.yml)


## Supported architectures
- linux/amd64
- linux/arm/v6
- linux/arm/v7
- linux/arm64
- linux/386


## Tested on
- Raspberry Pi 4B
- Raspberry Pi 3B
- Raspberry Pi Zero W (`dump1090` runs well, `fr24feed` is not particularly stable)


## Tested adapters
- *Should work with every RTL-SDR dongle*
- `0bda:2838` Realtek Semiconductor Corp. RTL2838 DVB-T


## Used software

### Encoders and more
- [readsb](https://github.com/wiedehopf/readsb)
- [dump1090](https://github.com/antirez/dump1090)
- [dump978](https://github.com/flightaware/dump978)

### Feeders
- [fr24feed](https://www.flightradar24.com/share-your-data)
- [adsb2influx](https://github.com/slintak/adsb2influx) (own fork, only for InfluxDB 2.0)
- [mlat-client](https://github.com/wiedehopf/mlat-client)

### Utilities
- [uat2esnt](https://github.com/adsbxchange/uat2esnt)

### Additional software
- [Alpine Linux](https://www.alpinelinux.org/)
- [OpenStreetMap](https://www.openstreetmap.org/)
- [Font Awesome](https://fontawesome.com/)
- [Leaflet](https://github.com/Leaflet/Leaflet)


## Run with docker-compose

### Usage

1. Download `docker-compose.dump1090.yml` or `docker-compose.readsb.yml`
2. Rename the downloaded file to `docker-compose.yml`
3. Edit `docker-compose.yml` and set services, ports and environment variables
4. ???
5. Profit!

### Commands

| Description      | Command                  |
|------------------|--------------------------|
| Start containers | `docker-compose up -d`   |
| Stop containers  | `docker-compose down`    |
| Pull images      | `docker-compose pull`    |
| See logs         | `docker-compose logs -f` |


## Image "readsb"
[![Docker Pulls](https://img.shields.io/docker/pulls/flighttracker/readsb) ![Docker Image Size (tag)](https://img.shields.io/docker/image-size/flighttracker/readsb/latest)](https://hub.docker.com/r/flighttracker/readsb)

### Environment variables

See `ENV`-lines in [Dockerfile](https://github.com/Dennis14e/docker-flighttracker/blob/main/readsb/Dockerfile).

### docker run

```
docker run \
  -it --rm \
  --device=/dev/bus/usb:/dev/bus/usb \
  flighttracker/readsb:latest
```


## Image "readsb-web"
[![Docker Pulls](https://img.shields.io/docker/pulls/flighttracker/readsb-web) ![Docker Image Size (tag)](https://img.shields.io/docker/image-size/flighttracker/readsb-web/latest)](https://hub.docker.com/r/flighttracker/readsb-web)

readsb-web is only compatible with readsb.

### Environment variables

| Environment variable | Default                 | Description             |
|----------------------|-------------------------|-------------------------|
| READSB_HOST          | readsb                  | readsb Host             |
| READSB_PORT          | 8042                    | readsb API Port         |
| PAGE_TITLE           | Flighttracker Web Light | Page title              |
| REFRESH_RATE         | 250                     | Data refresh rate in ms |
| CENTER_LAT           | 45.0                    | Map center latitude     |
| CENTER_LNG           | 9.0                     | Map center longitude    |
| CENTER_RANGE         | 500                     | Map range in nm         |

### docker run

```
docker run \
  -it --rm \
  -p 80:80 \
  -e CENTER_LAT=REPLACE \
  -e CENTER_LNG=REPLACE \
  -e CENTER_RANGE=REPLACE \
  flighttracker/readsb-web:latest
```


## Image "dump1090"
[![Docker Pulls](https://img.shields.io/docker/pulls/flighttracker/dump1090) ![Docker Image Size (tag)](https://img.shields.io/docker/image-size/flighttracker/dump1090/latest)](https://hub.docker.com/r/flighttracker/dump1090)

### Environment variables

| Environment variable | Default        | Description          |
|----------------------|----------------|----------------------|
| TITLE                | dump1090       | Page title           |
| GMAP_STYLE           | old            | Map style old/new    |
| GMAP_CENTER_LAT      | 45.0           | Map center latitude  |
| GMAP_CENTER_LNG      | 9.0            | Map center longitude |

### docker run

```
docker run \
  -it --rm \
  -p 8080:8080 \
  --device=/dev/bus/usb:/dev/bus/usb \
  -e GMAP_STYLE=new \
  -e GMAP_CENTER_LAT=45.0 \
  -e GMAP_CENTER_LNG=9.0 \
  flighttracker/dump1090:latest
```


## Image "dump978"
[![Docker Pulls](https://img.shields.io/docker/pulls/flighttracker/dump978) ![Docker Image Size (tag)](https://img.shields.io/docker/image-size/flighttracker/dump978/latest)](https://hub.docker.com/r/flighttracker/dump978)

### Warning

dump978 is currently not supported due to unexplained problems building Docker images.

### Environment variables

See `ENV`-lines in [Dockerfile](https://github.com/Dennis14e/docker-flighttracker/blob/main/dump978/Dockerfile).

### docker run

```
docker run \
  -it --rm \
  -p 30002:30002 \
  --device=/dev/bus/usb:/dev/bus/usb \
  flighttracker/dump978:latest
```


## Image "uat2esnt"
[![Docker Pulls](https://img.shields.io/docker/pulls/flighttracker/uat2esnt) ![Docker Image Size (tag)](https://img.shields.io/docker/image-size/flighttracker/uat2esnt/latest)](https://hub.docker.com/r/flighttracker/uat2esnt)

uat2esnt is only compatible with dump978.

### Environment variables

| Environment variable | Default | Description  |
|----------------------|---------|--------------|
| DUMP978_HOST         | dump978 | dump978 host |
| DUMP978_PORT         | 30002   | dump978 port |
| OUTPUT_PORT          | 30002   | Output port  |

### docker run

```
docker run \
  -it --rm \
  -p 30002:30002 \
  flighttracker/uat2esnt:latest
```


## Image "mlat-client"
[![Docker Pulls](https://img.shields.io/docker/pulls/flighttracker/mlat-client) ![Docker Image Size (tag)](https://img.shields.io/docker/image-size/flighttracker/mlat-client/latest)](https://hub.docker.com/r/flighttracker/mlat-client)

### Environment variables

| Environment variable | Default               | Description                                                                                                           |
|----------------------|-----------------------|-----------------------------------------------------------------------------------------------------------------------|
| INPUT_TYPE           | auto                  | Input type: `auto`, `dump1090`, `beast`, `radarcape_12mhz`, `radarcape_gps`, `radarcape`, `sbs`, `avrmlat`            |
| INPUT_CONNECT_HOST   |                       | Input host to connect to for Mode S traffic                                                                           |
| INPUT_CONNECT_PORT   | 30002                 | Input port to connect to for Mode S traffic                                                                           |
| RESULTS              |                       | `<protocol>,connect,<host:port>` or `<protocol>,listen,<port>`<br>Protocol: `basestation`, `ext_basestation`, `beast` |
| NO_ANON_RESULTS      | no                    | No results for anonymized aircraft: `yes`, `no`                                                                       |
| NO_MODEAC_RESULTS    | no                    | No results for Mode A/C tracks: `yes`, `no`                                                                           |
| LAT                  |                       | Latitude of receiver                                                                                                  |
| LON                  |                       | Longitude of receiver                                                                                                 |
| ALT                  |                       | Altitude of receiver (in `m` or `ft`)                                                                                 |
| PRIVACY              | no                    | Sets the privacy flag for this receiver: `yes`, `no`                                                                  |
| USER                 |                       | User information to give to the server                                                                                |
| SERVER_HOST          | feed.adsbexchange.com | Host of the multilateration server to connect to                                                                      |
| SERVER_PORT          | 31090                 | Port of the multilateration server to connect to                                                                      |
| NO_UDP               | no                    | Don't offer to use UDP transport for sync/mlat messages: `yes`, `no`                                                  |
| LOG_TIMESTAMPS       | no                    | Print timestamps in logging output: `yes`, `no`                                                                       |

### docker run
```
docker run \
  -it --rm \
  -e INPUT_CONNECT_HOST=readsb \
  -e INPUT_CONNECT_PORT=30002 \
  -e LAT= \
  -e LON= \
  -e ALT= \
  -e USER= \
  flighttracker/mlat-client:latest
```


## Image "fr24feed"
[![Docker Pulls](https://img.shields.io/docker/pulls/flighttracker/fr24feed) ![Docker Image Size (tag)](https://img.shields.io/docker/image-size/flighttracker/fr24feed/latest)](https://hub.docker.com/r/flighttracker/fr24feed)

fr24feed is compatible with dump1090 and readsb.

### Environment variables

| Environment variable | Default        | Description               |
|----------------------|----------------|---------------------------|
| SHARING_KEY          |                | Flightradar24 sharing key |
| DUMP1090_HOST        | dump1090:30002 | dump1090 Host:Port        |
| RECEIVER             | avr-tcp        | Receiver type             |
| BS                   | no             | BS                        |
| RAW                  | no             | RAW                       |
| LOGMODE              | 0              | Logging mode              |
| WINDOWMODE           | 0              | Window mode               |
| MPX                  | no             | MPX                       |
| MLAT                 | yes            | MLAT                      |
| MLAT_WITHOUT_GPS     | yes            | MLAT without GPS          |

### docker run

```
docker run \
  -it --rm \
  -p 8754:8754 \
  -e SHARING_KEY=REPLACE \
  -e DUMP1090_HOST=dump1090:30002 \
  flighttracker/fr24feed:latest
```


## Image "adsb2influx"
[![Docker Pulls](https://img.shields.io/docker/pulls/flighttracker/adsb2influx) ![Docker Image Size (tag)](https://img.shields.io/docker/image-size/flighttracker/adsb2influx/latest)](https://hub.docker.com/r/flighttracker/adsb2influx)

adsb2influx is compatible with dump1090 and readsb.

### Environment variables

| Environment variable | Default        | Description              |
|----------------------|----------------|--------------------------|
| DUMP1090_HOST        | dump1090       | dump1090 Host            |
| DUMP1090_PORT        | 30003          | dump1090 Port            |
| INFLUX_URL           |                | InfluxDB URL             |
| INFLUX_TOKEN         |                | InfluxDB API token       |
| INFLUX_ORG           |                | InfluxDB organisation    |
| INFLUX_BUCKET        | adsb           | InfluxDB bucket          |
| INFLUX_MEASUREMENT   | messages       | InfluxDB measurement     |
| SEND_INTERVAL        | 60             | Data send interval (sec) |

### docker run
```
docker run \
  -it --rm \
  -e DUMP1090_HOST=dump1090 \
  -e DUMP1090_PORT=30003 \
  -e INFLUX_URL= \
  -e INFLUX_TOKEN= \
  -e INFLUX_ORG= \
  -e INFLUX_BUCKET=adsb \
  -e SEND_INTERVAL=60 \
  flighttracker/adsb2influx:latest
```


CHANGELOG.md
# Changelog

## v1.0.2

- Update readsb
- Update fr24feed (arm)

## v1.0.1

- Update readsb
- Disable dump978 builds

## v1.0.0

- First release


adsb2influx/adsb2influx.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

# Imports
import signal
import argparse
import logging
import socket
import re
import time

from influxdb_client import InfluxDBClient, Point
from influxdb_client.client.write_api import SYNCHRONOUS
from influxdb_client.rest import ApiException


# Logging
logging.basicConfig(
    level = logging.INFO,
    format = '%(asctime)s - %(name)s - %(levelname)s - %(message)s'
)
log = logging.getLogger('adsb2influx')


# Killer
class GracefulKiller(object):
    kill_now = False

    def __init__(self):
        signal.signal(signal.SIGINT, self.handler)
        signal.signal(signal.SIGTERM, self.handler)

    def handler(self, sig, frame):
        self.kill_now = True


# AdsbProcessor
class AdsbError(Exception):
    pass

class AdsbProcessor(object):
    MSG_REGEX = r'^MSG,' \
        r'(?P<transmission>\d),' \
        r'(?P<session>\d*),' \
        r'(?P<aircraft>\d*),' \
        r'(?P<hexident>[0-9A-F]+),' \
        r'(?P<flight>\d*),' \
        r'(?P<gen_date>[0-9/]*),' \
        r'(?P<gen_time>[0-9:\.]*),' \
        r'(?P<log_date>[0-9/]*),' \
        r'(?P<log_time>[0-9:\.]*),' \
        r'(?P<callsign>[\w\s]*),' \
        r'(?P<altitude>\d*),' \
        r'(?P<speed>\d*),' \
        r'(?P<track>[\d\-]*),' \
        r'(?P<latitude>[\d\-\.]*),' \
        r'(?P<longitude>[\d\-\.]*),' \
        r'(?P<verticalrate>[\d\-]*),' \
        r'(?P<squawk>\d*),' \
        r'(?P<alert>[\d\-]*),' \
        r'(?P<emergency>[\d\-]*),' \
        r'(?P<spi>[\d\-]*),' \
        r'(?P<onground>[\d\-]*)$'

    MSG_NORMAL = {
        'transmission': (lambda v: int(v)),
        'session': (lambda v: int(v)),
        'aircraft': (lambda v: int(v)),
        'flight': (lambda v: int(v)),
        'callsign': (lambda v: v.strip()),
        'altitude': (lambda v: int(v)),
        'speed': (lambda v: int(v)),
        'track': (lambda v: int(v)),
        'latitude': (lambda v: float(v)),
        'longitude': (lambda v: float(v)),
        'verticalrate': (lambda v: int(v)),
        'alert': (lambda v: True if v == '-1' else False),
        'emergency': (lambda v: True if v == '-1' else False),
        'spi': (lambda v: True if v == '-1' else False),
        'onground': (lambda v: True if v == '-1' else False),
    }

    def __init__(self):
        self.re_msg = re.compile(self.MSG_REGEX)
        self.aircrafts = {}
        self.aircrafts_age = {}

    def __getitem__(self, key):
        return self.aircrafts[key]

    def __setitem__(self, key, value):
        self.aircrafts[key] = value

    def __delitem__(self, key):
        del self.aircrafts[key]

    def __contains__(self, key):
        return key in self.aircrafts

    def __len__(self):
        return len(self.aircrafts)

    def __repr__(self):
        return repr(self.aircrafts)

    def __cmp__(self, dict_):
        return self.__cmp__(self.aircrafts, dict_)

    def __iter__(self):
        return iter(self.aircrafts)

    def __unicode__(self):
        return unicode(repr(self.aircrafts))

    def __normalize_msg(self, msg):
        for field, fnc in self.MSG_NORMAL.items():
            if field in msg:
                msg[field] = fnc(msg[field])

        return msg

    def keys(self):
        return self.aircrafts.keys()

    def values(self):
        return self.aircrafts.values()

    def items(self):
        return self.aircrafts.items()

    def pop(self, *args):
        return self.aircrafts.pop(*args)

    def clear(self, age):
        for hexident in list(self.aircrafts_age.keys()):
            if hexident in self.aircrafts:
                del self.aircrafts_age[hexident]
                continue

            if (time.time() - self.aircrafts_age[hexident]) > age:
                log.info('Hexident {} is too old, deleting'.format(hexident))
                del self.aircrafts_age[hexident]
                del self.aircrafts[hexident]

        for hexident in self.aircrafts.keys():
            self.aircrafts[hexident]['count'] = 0

    def age(self, hexident):
        return (time.time() - self.aircrafts_age.get(hexident, 0))

    def msg(self, data):
        data = data.strip()
        log.debug(data)

        matches = self.re_msg.match(data)
        if not matches:
            log.error('Wrong format for MSG \'{}\', skipping...'.format(data))
            return

        msg = { k: v for k, v in matches.groupdict().items() if v }
        msg = self.__normalize_msg(msg)

        self.aircrafts_age[msg['hexident']] = time.time()

        if msg['hexident'] not in self.aircrafts:
            self.aircrafts[msg['hexident']] = msg
            self.aircrafts[msg['hexident']]['count'] = 1
        else:
            self.aircrafts[msg['hexident']].update(msg)
            self.aircrafts[msg['hexident']]['count'] += 1


# dump1090
class Dump1090(object):
    def __init__(self, host, port):
        self.host = host
        self.port = port
        self.socket = None
        self.data = ''

    def connect(self):
        log.info('Connecting to dump1090 on {}:{}'.format(self.host, self.port))
        self.socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        connected = False

        while not connected:
            try:
                self.socket.connect((self.host, self.port))
                connected = True
                log.info('Connected OK, receiving data')
            except Exception as e:
                connected = False
                log.warning('Could not connect, retrying ({})'.format(e))

        self.socket.setblocking(False)
        self.socket.settimeout(1)

    def disconnect(self):
        self.socket.close()
        log.info('Disconnected from dump1090')

    def receive(self):
        ret = None

        try:
            self.data += self.socket.recv(1024).decode('UTF-8')
            self.socket.send(bytes("\n", 'UTF-8'))

            newline = self.data.find('\n')
            if newline >= 0:
                ret = self.data[:newline]
                self.data = self.data[newline + 1:]
        except socket.timeout:
            pass
        except Exception as e:
            raise AdsbError('Error receiving data from dump1090: \'{}\''.format(e))

        return ret


# InfluxDB
class InfluxDB(object):
    def __init__(self, url, token, org, bucket):
        self.org = org
        self.bucket = bucket

        self.client = InfluxDBClient(
            url = url,
            token = token,
            org = org
        )

        self.writeapi = self.client.write_api(
            write_options = SYNCHRONOUS
        )

    def write(self, data):
        log.debug('Write data to InfluxDB: {}'.format(data))

        try:
            self.writeapi.write(self.bucket, self.org, data)
            log.info('Written {} aircraft to InfluxDB'.format(len(data)))
        except ApiException as e:
            log.error('Writing data to InfluxDB failed, status code: {} {}'.format(e.status, e.reason))
            return False

        return True


# Main
def main():
    parser = argparse.ArgumentParser(
        description = 'Write ADSB data from dump1090 to InfluxDB 2.0'
    )

    parser.add_argument(
        '-dh', '--dump1090-host',
        dest = 'dump1090_host',
        default = 'localhost',
        help = 'dump1090 host/ip (default: localhost)'
    )
    parser.add_argument(
        '-dp', '--dump1090-port',
        dest = 'dump1090_port',
        type = int,
        default = 30003,
        help = 'dump1090 port (default: 30003)'
    )
    parser.add_argument(
        '-iu', '--influxdb-url',
        dest = 'influxdb_url',
        required = True,
        help = 'InfluxDB url (e.g. http://url-to-influxdb:8086/)'
    )
    parser.add_argument(
        '-it', '--influxdb-token',
        dest = 'influxdb_token',
        required = True,
        help = 'InfluxDB API token'
    )
    parser.add_argument(
        '-io', '--influxdb-org',
        dest = 'influxdb_org',
        required = True,
        help = 'InfluxDB organisation'
    )
    parser.add_argument(
        '-ib', '--influxdb-bucket',
        dest = 'influxdb_bucket',
        default = 'adsb',
        help = 'InfluxDB bucket (default: adsb)'
    )
    parser.add_argument(
        '-im', '--influxdb-measurement',
        dest = 'influxdb_measurement',
        default = 'messages',
        help = 'InfluxDB measurement (default: messages)'
    )
    parser.add_argument(
        '-si', '--send-interval',
        dest = 'send_interval',
        type = int,
        default = 60,
        help = 'send interval in seconds (default: 60)'
    )
    parser.add_argument(
        '-d', '--debug',
        dest = 'debug',
        action = 'store_true',
        help = 'set logging to debug level'
    )

    args = parser.parse_args()


    # Logging
    if args.debug == True:
        log.setLevel(logging.DEBUG)

    log.debug('Arguments: {}'.format(args))


    # Killer
    killer = GracefulKiller()

    # AdsbProcessor
    ap = AdsbProcessor()

    # dump1090
    dump1090 = Dump1090(args.dump1090_host, args.dump1090_port)
    dump1090.connect()

    # InfluxDB
    influxdb = InfluxDB(
        url = args.influxdb_url,
        token = args.influxdb_token,
        org = args.influxdb_org,
        bucket = args.influxdb_bucket
    )


    measurement = args.influxdb_measurement
    send_interval = args.send_interval
    last_print = time.time()


    # This is where the magic happens
    while not killer.kill_now:
        if (time.time() - last_print) > send_interval:
            last_print = time.time()

            to_send = []

            for hexident, msg in ap.items():
                if not all(k in msg for k in ['callsign', 'squawk']):
                    log.info('Missing callsign or squawk for {}'.format(hexident))
                    continue

                if ap.age(hexident) > send_interval:
                    log.info('Aircraft {} was not seen too long. Not sending'.format(hexident))
                    continue

                to_send.append({
                    'measurement': measurement,
                    'tags': {
                        'hexident': hexident,
                        'callsign': msg['callsign'],
                        'squawk': msg['squawk'],
                    },
                    'fields': {
                        'generated': time.time(),
                        'altitude': msg.get('altitude'),
                        'speed': msg.get('speed'),
                        'track': msg.get('track'),
                        'latitude': msg.get('latitude'),
                        'longitude': msg.get('longitude'),
                        'verticalrate': msg.get('verticalrate'),
                        'alert': msg.get('alert'),
                        'emergency': msg.get('emergency'),
                        'spi': msg.get('spi'),
                        'onground': msg.get('onground'),
                        'count': msg.get('count', 0),
                    }
                })


            ap.clear(send_interval * 3)
            if len(to_send) > 0:
                influxdb.write(to_send)
            else:
                log.info('No aircrafts to be saved in InfluxDB')


        try:
            msg = dump1090.receive()
        except AdsbError as e:
            log.error(e)
            killer.kill_now = True
        else:
            if msg is not None:
                ap.msg(msg)


    # Exit
    dump1090.disconnect()


if __name__ == "__main__":
    main()
